src/libos/crates/async-rt/src/sync/mutex/inner.rs

```diff
@@ -0,0 +1,221 @@
+    waiter_queue: Mutex<WaiterQueue>,
+}
+
+// This struct is the atomic wrapper for LockStatus.
+#[derive(Debug)]
+struct AtomicLockStatus(AtomicI32);
+
+// The status code keeps the same as the implementation in musl.
+#[derive(Debug, Copy, Clone)]
+#[repr(i32)]
+enum LockStatus {
+    Free = 0,
+    Locked = Errno::EBUSY as i32,
+    LockedWithWaiters = -2147483632, // 0x8000_0010 = (EBUSY | 0x8000_0000)
+}
+
+impl MutexInner {
+    pub(super) fn new() -> MutexInner {
+        MutexInner {
+            status: AtomicLockStatus::init(),
+            waiters: AtomicI32::new(0),
+            waiter_queue: Mutex::new(WaiterQueue::new()),
+        }
+    }
+
+    pub(super) fn try_lock(&self) -> Result<()> {
+        if self.status.try_set_lock().is_ok() {
+            Ok(())
+        } else {
+            Err(errno!(EBUSY, "the lock is held by other threads"))
+        }
+    }
+
+    pub(super) async fn lock(&self) -> Result<()> {
+        if let Ok(_) = self.try_lock() {
+            return Ok(());
+        }
+
+        // Spin for a short while if no one is waiting but the lock is held.
+        let mut spins = SPIN_COUNT;
+        while spins != 0
+            && self.status.is_locked()
+            // Can't reorder here. `Relaxed` is enough.
+            && self.waiters.load(Ordering::Relaxed) == 0
+        {
+            hint::spin_loop();
+            spins -= 1;
+        }
+
+        loop {
+            if let Ok(_) = self.try_lock() {
+                return Ok(());
+            }
+
+            if self.status.is_free() {
+                continue;
+            }
+
+            // In try_set_lock_with_waiters, `AcqRel` will make sure this increment happens before. Thus, `Relaxed` can be used here.
+            self.waiters.fetch_add(1, Ordering::Relaxed);
+
+            // Ignore the result here. If the state transition fails, the next wait will not block.
+            self.status
+                .try_set_lock_with_waiters()
+                .map_err(|e| errno!(e.errno(), "failed to set lock status"))
+                .ok();
+
+            let ret = self.wait().await;
+            // Use `Acquire` to make sure `ret` is already set.
+            self.waiters.fetch_sub(1, Ordering::Acquire);
+
+            if let Err(error) = &ret {
+                match error.errno() {
+                    Errno::EINTR => continue,
+                    _ => {
+                        return_errno!(error.errno(), "futex wait error")
+                    }
+                }
+            }
+        }
+    }
+
+    pub(super) fn unlock(&self) -> Result<()> {
+        // `get_current_status_and_set_free` will make sure this happens before. Thus, `Relaxed` can be used here.
+        let waiters = self.waiters.load(Ordering::Relaxed);
+        let previous_status = self.status.get_current_status_and_set_free();
+
+        // Also check previous status in case the waiter number overflows.
+        if waiters > 0 || previous_status.is_locked_with_waiters() {
```

Comment:
Why checking `is_locked_with_waiters` is going to help with overflow? Could you explain?

Reply:
I didn't get it. You are basically saying that having `waiters` does not really matter (because it can overflow anyway) and all we need to do is to check `is_locked_with_waiters`. If so, why bother keeping the waiters counter? We can just check `is_locked_with_waiters` alone.

Reply:
The `waiters` checking is the main check. `is_locked_with_waiters` is only to protect the situation that the waiter counter overflows to 0. 
`is_locked_with_waiters` can't be used as the main check because the status will be `Locked` when thread 1 `unlock` and `wake` up thread 2 and thread 2 `try_lock` succeeds. Thus, when thread 2 `unlock`, the `is_locked_with_waiters` check will fail and other waiters will not be woken up. 
However, `is_locked_with_waiters` check can guard the situation that the waiter counter overflows to 0 because this situation only exists once. And in the next unlock, the waiter counter will not be 0 anymore.